Unreadable exception on serialization error

Hi Guys,

I found that it's really hard to get what's happening if we getting a serialization error during execution of the CPS tests:
```
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.226 sec <<< FAILURE!
default_run(CPS1Test)  Time elapsed: 0.79 sec  <<< ERROR!
java.lang.Exception: com.cloudbees.groovy.cps.impl.BlockScopeEnv@7393222f
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
...
Caused by: java.io.NotSerializableException: org.example.testsharedlib.TestClass
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at java.util.HashMap.internalWriteEntries(HashMap.java:1790)
	at java.util.HashMap.writeObject(HashMap.java:1363)
	at sun.reflect.GeneratedMethodAccessor10.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
...
```

Of course it contains the caused by exception - but many ones can't find it and `com.cloudbees.groovy.cps.impl.BlockScopeEnv@7393222f` is not helping to locate the place of exception.

Maybe it will be better to slightly modify the PipelineHelperCPS to get the next result:
```
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.116 sec <<< FAILURE!
default_run(CPS2Test)  Time elapsed: 0.75 sec  <<< ERROR!
java.lang.Exception: Unable to serialize locals [test_obj:org.example.testsharedlib.TestClass@6d167f58] in test
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
...
Caused by: java.io.NotSerializableException: org.example.testsharedlib.TestClass
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
...
```

Looks much better and more clean for the end user. Or at least remove the try to capture the exception - it makes no sense.

Context shared library attached: [test_shared_lib-JenkinsPipelineUnit-160.zip](https://github.com/jenkinsci/JenkinsPipelineUnit/files/3812182/test_shared_lib-JenkinsPipelineUnit-160.zip)

--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,4 @@ build
 .gradle
 *.swp
 out
+.DS_store

--- a/src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy
+++ b/src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy
@@ -74,6 +74,7 @@ abstract class BasePipelineTest {
         helper.registerAllowedMethod('archive', [String.class], null)
         helper.registerAllowedMethod("archiveArtifacts", [Map.class], null)
         helper.registerAllowedMethod('archiveArtifacts', [String.class], null)
+        helper.registerAllowedMethod("bat", [String.class], null)
         helper.registerAllowedMethod("build", [Map.class], {
             [
                 getNumber:{100500},

--- a/src/main/groovy/com/lesfurets/jenkins/unit/cps/PipelineTestHelperCPS.groovy
+++ b/src/main/groovy/com/lesfurets/jenkins/unit/cps/PipelineTestHelperCPS.groovy
@@ -74,6 +74,11 @@ class PipelineTestHelperCPS extends PipelineTestHelper {
             while(next.yield==null) {
                 try {
                     this.roundtripSerialization(next.e)
+                } catch (NotSerializableException exception) {
+                  def cls = exception.getMessage()
+                  def vars = next.e.locals.findAll { k, v -> v.class == cls }
+                  vars = vars.isEmpty() ? 'UNKNOWN' : vars
+                  throw new RuntimeException("Unable to serialize locals ${vars} in `${next.e.closureOwner().class}`", exception)
                 } catch (exception) {
                     throw new Exception(next.e.toString(), exception)
                 }

--- a/src/test/groovy/com/lesfurets/jenkins/TestNotSerrialzibleCPS.groovy
+++ b/src/test/groovy/com/lesfurets/jenkins/TestNotSerrialzibleCPS.groovy
@@ -1,0 +1,46 @@
+package com.lesfurets.jenkins
+
+import org.junit.Before
+import org.junit.Test
+import org.junit.Rule
+import org.junit.rules.ExpectedException
+
+import com.lesfurets.jenkins.unit.cps.BasePipelineTestCPS
+import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
+import static com.lesfurets.jenkins.unit.global.lib.ProjectSource.projectSource
+
+class TestNotSerrialzibleCPS extends BasePipelineTestCPS {
+    // @Rule
+    // public ExpectedException thrown = ExpectedException.none();
+
+    String sharedLibs = this.class
+        .getResource('/libs/test_not_serrializable_cps')
+        .getFile()
+
+    @Override
+    @Before
+    void setUp() throws Exception {
+        helper.registerSharedLibrary(library()
+            .name('test_not_serrializable_cps')
+            .allowOverride(false)
+            .retriever(projectSource(sharedLibs))
+            .targetPath(sharedLibs)
+            .defaultVersion('master')
+            .implicit(true)
+            .build()
+        )
+        scriptRoots += 'src/test/jenkins'
+        super.setUp()
+
+    }
+
+    @Test
+    void default_run() throws Exception {
+        //thrown.expect(RuntimeException.class)
+        //thrown.expectMessage(containsString("Unable to serialize locals"))
+        runScript('job/testNotSerrializableCPS.jenkins')
+        printCallStack()
+    }
+
+
+}

--- a/src/test/groovy/com/lesfurets/jenkins/TestNotSerrialzibleCPS.groovy
+++ b/src/test/groovy/com/lesfurets/jenkins/TestNotSerrialzibleCPS.groovy
@@ -10,8 +10,8 @@ import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
 import static com.lesfurets.jenkins.unit.global.lib.ProjectSource.projectSource
 
 class TestNotSerrialzibleCPS extends BasePipelineTestCPS {
-    // @Rule
-    // public ExpectedException thrown = ExpectedException.none();
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
 
     String sharedLibs = this.class
         .getResource('/libs/test_not_serrializable_cps')
@@ -35,10 +35,11 @@ class TestNotSerrialzibleCPS extends BasePipelineTestCPS {
     }
 
     @Test
-    void default_run() throws Exception {
-        //thrown.expect(RuntimeException.class)
-        //thrown.expectMessage(containsString("Unable to serialize locals"))
-        runScript('job/testNotSerrializableCPS.jenkins')
+    void default_run() {
+        thrown.expect(Exception.class)
+        thrown.expectMessage(containsString("Unable to serialize locals"))
+        def script = loadScript('job/testNotSerrializableCPS.jenkins')
+        script.execute()
         printCallStack()
     }
 

--- a/src/test/jenkins/job/testNotSerrializableCPS.jenkins
+++ b/src/test/jenkins/job/testNotSerrializableCPS.jenkins
@@ -1,0 +1,4 @@
+def execute() {
+    def a = 3
+    echo "Test Step: ${test()}"
+}

--- a/src/test/resources/libs/test_not_serrializable_cps/src/org/example/testsharedlib/TestClass.groovy
+++ b/src/test/resources/libs/test_not_serrializable_cps/src/org/example/testsharedlib/TestClass.groovy
@@ -1,0 +1,7 @@
+package org.example.testsharedlib
+
+public class TestClass {
+  String test1() {
+    return 'ok'
+  }
+}

--- a/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/CPS1Test.groovy
+++ b/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/CPS1Test.groovy
@@ -1,0 +1,39 @@
+import org.junit.Before
+import org.junit.Test
+
+import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
+import static com.lesfurets.jenkins.unit.global.lib.LocalSource.localSource
+
+import static org.assertj.core.api.Assertions.assertThat
+
+import com.lesfurets.jenkins.unit.cps.BasePipelineTestCPS
+
+class CPS1Test extends BasePipelineTestCPS {
+  @Override
+  @Before
+  void setUp() throws Exception {
+    String sharedLibs = this.class.getResource('.').getFile()
+
+    helper.registerSharedLibrary(library()
+        .name('testsharedlib')
+        .allowOverride(false)
+        .retriever(localSource(sharedLibs))
+        .targetPath(sharedLibs)
+        .defaultVersion('snapshot')
+        .implicit(true)
+        .build()
+    )
+
+    setScriptRoots([ 'jobs' ] as String[])
+    setScriptExtension('groovy')
+
+    super.setUp()
+  }
+
+  @Test
+  void default_run() throws Exception {
+    def script = loadScript('exampleJob.jenkins')
+    script.execute()
+    printCallStack()
+  }
+}

--- a/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/CPS2Test.groovy
+++ b/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/CPS2Test.groovy
@@ -1,0 +1,39 @@
+import org.junit.Before
+import org.junit.Test
+
+import static com.lesfurets.jenkins.unit.global.lib.LibraryConfiguration.library
+import static com.lesfurets.jenkins.unit.global.lib.LocalSource.localSource
+
+import static org.assertj.core.api.Assertions.assertThat
+
+import org.example.testsharedlib.ModBasePipelineTestCPS
+
+class CPS2Test extends ModBasePipelineTestCPS {
+  @Override
+  @Before
+  void setUp() throws Exception {
+    String sharedLibs = this.class.getResource('.').getFile()
+
+    helper.registerSharedLibrary(library()
+        .name('testsharedlib')
+        .allowOverride(false)
+        .retriever(localSource(sharedLibs))
+        .targetPath(sharedLibs)
+        .defaultVersion('snapshot')
+        .implicit(true)
+        .build()
+    )
+
+    setScriptRoots([ 'jobs' ] as String[])
+    setScriptExtension('groovy')
+
+    super.setUp()
+  }
+
+  @Test
+  void default_run() throws Exception {
+    def script = loadScript('exampleJob.jenkins')
+    script.execute()
+    printCallStack()
+  }
+}

--- a/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/ModBasePipelineTestCPS.groovy
+++ b/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/ModBasePipelineTestCPS.groovy
@@ -1,0 +1,11 @@
+package org.example.testsharedlib
+
+import com.lesfurets.jenkins.unit.cps.BasePipelineTestCPS
+
+import org.example.testsharedlib.ModPipelineTestHelperCPS
+
+abstract class ModBasePipelineTestCPS extends BasePipelineTestCPS {
+  ModBasePipelineTestCPS() {
+    helper = new ModPipelineTestHelperCPS()
+  }
+}

--- a/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/ModPipelineTestHelperCPS.groovy
+++ b/src/test/resources/libs/test_not_serrializable_cps/test/groovy/org/example/testsharedlib/ModPipelineTestHelperCPS.groovy
@@ -1,0 +1,46 @@
+package org.example.testsharedlib
+
+import com.lesfurets.jenkins.unit.cps.PipelineTestHelperCPS
+import com.lesfurets.jenkins.unit.PipelineTestHelper
+
+import com.cloudbees.groovy.cps.impl.CpsCallableInvocation
+import com.cloudbees.groovy.cps.Envs
+import com.cloudbees.groovy.cps.Continuation
+
+@groovy.transform.InheritConstructors
+class ModPipelineTestHelperCPS extends PipelineTestHelperCPS {
+  @Override
+  Object callClosure(Closure closure, Object[] args = null) {
+      try {
+          // MOD: Just to make sure we will execute PipelineTestHelper method
+          this.callClosure2(closure, args)
+      } catch(CpsCallableInvocation e) {
+          def next = e.invoke(Envs.empty(), null, Continuation.HALT)
+          while(next.yield==null) {
+              // MOD: Removed custom exception to show better info
+              try {
+                  this.roundtripSerialization(next.e)
+              } catch (NotSerializableException exception) {
+                  def cls = exception.getMessage()
+                  def vars = next.e.locals.findAll { k, v -> v.class.name == cls }
+                  vars = vars.isEmpty() ? 'UNKNOWN' : vars
+                  throw new NotSerializableException("Unable to serialize locals ${vars} in `${next.e.closureOwner().class.name}`", exception)
+              } catch (exception) {
+                  throw new Exception("Exception during serialization of locals ${next.e.locals} in `${next.e.closureOwner().class.name}`", exception)
+              }
+              next = next.step()
+          }
+          return next.yield.replay()
+      }
+  }
+
+  Object callClosure2(Closure closure, Object[] args = null) {
+      if (!args) {
+          return closure.call()
+      } else if (args.size() > closure.maximumNumberOfParameters) {
+          return closure.call(args)
+      } else {
+          return closure.call(*args)
+      }
+  }
+}

--- a/src/test/resources/libs/test_not_serrializable_cps/vars/test.groovy
+++ b/src/test/resources/libs/test_not_serrializable_cps/vars/test.groovy
@@ -1,0 +1,8 @@
+import org.example.testsharedlib.TestClass
+
+def call() {
+    def b = 4
+    def test_obj = new TestClass()
+    def g = 8
+    return "${test_obj.test1()}"
+}

